File: w3ggo/detect_retraining.go
package w3ggo

const detectionTimeRange = 60 * 1000

// getRetrainingIndex mirrors getRetrainingIndex from detectRetraining.ts.
// Returns the index in abilityOrder where a retraining marker should be inserted,
// or -1 if not detected.
func getRetrainingIndex(abilityOrder []AbilityOrderEntry, timeOfTomePurchase int) int {
	if len(abilityOrder) < 3 {
		return -1
	}

	candidate := &abilityOrder[0]
	candidateIndex := 0
	abilitiesInRange := 0

	for i := 1; i < len(abilityOrder); i++ {
		entry := &abilityOrder[i]
		diff := entry.Time - candidate.Time
		if diff < 0 {
			diff = -diff
		}
		if diff < detectionTimeRange {
			abilitiesInRange++
		} else {
			abilitiesInRange = 0
			candidate = entry
			candidateIndex = i
		}

		candidateDiff := candidate.Time - timeOfTomePurchase
		if candidateDiff < 0 {
			candidateDiff = -candidateDiff
		}
		if abilitiesInRange == 2 && candidateDiff <= detectionTimeRange {
			return candidateIndex
		}
	}
	return -1
}

File: w3grs/src/detect_retraining.rs
use crate::player::AbilityOrderEntry;

const DETECTION_TIME_RANGE: u32 = 60 * 1000;

/// Mirrors getRetrainingIndex from detectRetraining.ts.
/// Returns the index in ability_order where a retraining marker should be inserted,
/// or -1 if not detected.
pub fn get_retraining_index(
    ability_order: &[AbilityOrderEntry],
    time_of_tome_purchase: u32,
) -> i32 {
    if ability_order.len() < 3 {
        return -1;
    }

    let mut candidate = &ability_order[0];
    let mut candidate_index: i32 = 0;
    let mut abilities_in_range: usize = 0;

    for i in 1..ability_order.len() {
        let entry = &ability_order[i];
        if entry.time().saturating_sub(candidate.time()) < DETECTION_TIME_RANGE {
            abilities_in_range += 1;
        } else {
            abilities_in_range = 0;
            candidate = entry;
            candidate_index = i as i32;
        }

        if abilities_in_range == 2
            && candidate.time().saturating_sub(time_of_tome_purchase) <= DETECTION_TIME_RANGE
        {
            return candidate_index;
        }
    }
    -1
}
